File: 2048-iOS-Objc-C/m2048/Controller/M2ViewController.m
//
//  M2ViewController.m
//  m2048
//
//  Created by Danqing on 3/16/14.
//  Copyright (c) 2014 Danqing. All rights reserved.
//

#import "M2ViewController.h"

#import "M2Scene.h"
#import "M2GameManager.h"
#import "M2ScoreView.h"
#import "M2Overlay.h"
#import "M2GridView.h"

#import <Skillz/Skillz.h>

@implementation M2ViewController {
  IBOutlet UIButton *_forfeitButton;
  IBOutlet UILabel *_targetScore;
  IBOutlet UILabel *_subtitle;
  IBOutlet M2ScoreView *_scoreView;
  IBOutlet M2ScoreView *_bestView;
  
  M2Scene *_scene;
  
  IBOutlet M2Overlay *_overlay;
  IBOutlet UIImageView *_overlayBackground;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    [self updateState];

    _bestView.score.text = [NSString stringWithFormat:@"%ld", (long)[Settings integerForKey:@"Best Score"]];

    _forfeitButton.layer.cornerRadius = [GSTATE cornerRadius];
    _forfeitButton.layer.masksToBounds = YES;

    _overlay.hidden = YES;
    _overlayBackground.hidden = YES;

    // Configure the view.
    SKView * skView = (SKView *)self.view;

    // Create and configure the scene.
    M2Scene * scene = [M2Scene sceneWithSize:skView.bounds.size];
    scene.scaleMode = SKSceneScaleModeAspectFill;

    // Present the scene.
    [skView presentScene:scene];
    [self updateScore:0];

    _scene = scene;
    _scene.controller = self;
    
    [[Skillz skillzInstance] launchSkillz];
}


- (void)updateState
{
  [_scoreView updateAppearance];
  [_bestView updateAppearance];
  
  _forfeitButton.backgroundColor = [GSTATE buttonColor];
  _forfeitButton.titleLabel.font = [UIFont fontWithName:[GSTATE boldFontName] size:14];
   
  _targetScore.textColor = [GSTATE buttonColor];
  
  long target = [GSTATE valueForLevel:GSTATE.winningLevel];
  
  if (target > 100000) {
    _targetScore.font = [UIFont fontWithName:[GSTATE boldFontName] size:34];
  } else if (target < 10000) {
    _targetScore.font = [UIFont fontWithName:[GSTATE boldFontName] size:42];
  } else {
    _targetScore.font = [UIFont fontWithName:[GSTATE boldFontName] size:40];
  }
  
  _targetScore.text = [NSString stringWithFormat:@"%ld", target];
  
  _subtitle.textColor = [GSTATE buttonColor];
  _subtitle.font = [UIFont fontWithName:[GSTATE regularFontName] size:14];
  _subtitle.text = [NSString stringWithFormat:@"Join the numbers to get to %ld!", target];
  
  _overlay.message.font = [UIFont fontWithName:[GSTATE boldFontName] size:36];
  _overlay.keepPlaying.titleLabel.font = [UIFont fontWithName:[GSTATE boldFontName] size:17];
  _overlay.restartGame.titleLabel.font = [UIFont fontWithName:[GSTATE boldFontName] size:17];
  
  _overlay.message.textColor = [GSTATE buttonColor];
  [_overlay.keepPlaying setTitleColor:[GSTATE buttonColor] forState:UIControlStateNormal];
  [_overlay.restartGame setTitleColor:[GSTATE buttonColor] forState:UIControlStateNormal];
}


- (void)updateScore:(NSInteger)score
{
    _scoreView.score.text = [NSString stringWithFormat:@"%ld", (long)score];
    if ([Settings integerForKey:@"Best Score"] < score) {
        [Settings setInteger:score forKey:@"Best Score"];
        _bestView.score.text = [NSString stringWithFormat:@"%ld", (long)score];
    }
    
    if ([[Skillz skillzInstance] tournamentIsInProgress]) {
        [[Skillz skillzInstance] updatePlayersCurrentScore:@(score)];
    }
}


- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
  // Pause Sprite Kit. Otherwise the dismissal of the modal view would lag.
  ((SKView *)self.view).paused = YES;
}


- (IBAction)forfeitMatch:(id)sender
{
    [self hideOverlay];
    [self updateScore:0];
    
    if (![[Skillz skillzInstance] tournamentIsInProgress]) {
        [self->_scene startNewGame];
        return;
    }
    
    UIAlertController *abortAlertController = [UIAlertController alertControllerWithTitle:@"Forfeit Match?"
                                                                 message:@"A match is in progress. Are you sure you want to forfeit the match?"
                                                                 preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *yesAction = [UIAlertAction actionWithTitle:@"Yes"
                                              style:UIAlertActionStyleDefault
                                              handler:^(UIAlertAction* action)
    {
        [[Skillz skillzInstance] notifyPlayerAbortWithCompletion:^(void)
        {
            NSLog(@"The user forfeited the match.");
        }];
    }];
    
    UIAlertAction *noAction = [UIAlertAction actionWithTitle:@"No" style:UIAlertActionStyleCancel handler:^(UIAlertAction* action)
    {
        NSLog(@"Forfeit canceled, continuing the current match.");
    }];
    
    [abortAlertController addAction:yesAction];
    [abortAlertController addAction:noAction];
    [self presentViewController:abortAlertController animated:YES completion:nil];
}


- (IBAction)keepPlaying:(id)sender
{
  [self hideOverlay];
}


- (IBAction)done:(UIStoryboardSegue *)segue
{
  ((SKView *)self.view).paused = NO;
  if (GSTATE.needRefresh) {
    [GSTATE loadGlobalState];
    [self updateState];
    [self updateScore:0];
    [_scene startNewGame];
  }
}

- (void)startNewGame
{
    [_scene startNewGame];
}

- (void)endGame:(BOOL)won
{
  _overlay.hidden = NO;
  _overlay.alpha = 0;
  _overlayBackground.hidden = NO;
  _overlayBackground.alpha = 0;
  
  if (!won) {
    _overlay.keepPlaying.hidden = YES;
    _overlay.message.text = @"Game Over";
  } else {
    _overlay.keepPlaying.hidden = NO;
    _overlay.message.text = @"You Win!";
  }
  
  // Fake the overlay background as a mask on the board.
  _overlayBackground.image = [M2GridView gridImageWithOverlay];
  
  // Center the overlay in the board.
  CGFloat verticalOffset = [[UIScreen mainScreen] bounds].size.height - GSTATE.verticalOffset;
  NSInteger side = GSTATE.dimension * (GSTATE.tileSize + GSTATE.borderWidth) + GSTATE.borderWidth;
  _overlay.center = CGPointMake(GSTATE.horizontalOffset + side / 2, verticalOffset - side / 2);
  
  [UIView animateWithDuration:0.5 delay:1.5 options:UIViewAnimationOptionCurveEaseInOut animations:^{
      self->_overlay.alpha = 1;
      self->_overlayBackground.alpha = 1;
  } completion:^(BOOL finished) {
    // Freeze the current game.
    ((SKView *)self.view).paused = YES;
      
      if ([[Skillz skillzInstance] tournamentIsInProgress]) {
          NSInteger matchId = [[[Skillz skillzInstance] getMatchInfo] id];
          
          [[Skillz skillzInstance] displayTournamentResultsWithScore:@([self->_scoreView.score.text integerValue])
                                                         withMatchId:@(matchId)
                                                      withCompletion:^(void)
          {
              NSLog(@"The match has ended, and the user's final score was reported.");
          }];
      }
  }];
}


- (void)hideOverlay
{
  ((SKView *)self.view).paused = NO;
  if (!_overlay.hidden) {
    [UIView animateWithDuration:0.5 animations:^{
        self->_overlay.alpha = 0;
        self->_overlayBackground.alpha = 0;
    } completion:^(BOOL finished) {
        self->_overlay.hidden = YES;
        self->_overlayBackground.hidden = YES;
    }];
  }
}


- (void)didReceiveMemoryWarning
{
  [super didReceiveMemoryWarning];
  // Release any cached data, images, etc that aren't in use.
}

@end

File: 2048-iOS-Swift/m2048/Controller/M2ViewController.swift
//
//  M2ViewController.swift
//  m2048
//
//  Created by Skillz on 5/12/20.
//  Copyright © 2020 Danqing. All rights reserved.
//

import UIKit
import SpriteKit
import Skillz

class M2ViewController : UIViewController {
    @IBOutlet var  _forfeitButton: UIButton?
    @IBOutlet var _targetScore: UILabel?
    @IBOutlet var _subtitle: UILabel?
    @IBOutlet var _scoreView: M2ScoreView?
    @IBOutlet var _bestView: M2ScoreView?

    @IBOutlet var  _overlay: M2Overlay?
    @IBOutlet var  _overlayBackground: UIImageView?
    
    var _scene: M2Scene?

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        self._scene = nil
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }

    required init?(coder: NSCoder) {
        self._scene = nil
        super.init(coder: coder)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.updateState()

        _bestView?.score?.text = String(format: "%ld", Settings.integer(forKey: "Best Score"))

        _forfeitButton?.layer.cornerRadius = CGFloat(GSTATE.cornerRadius)
        _forfeitButton?.layer.masksToBounds = true;

        _overlay?.isHidden = true;
        _overlayBackground?.isHidden = true;

        // Configure the view.
        let skView = self.view as! SKView

        // Create and configure the scene.
        let scene = M2Scene(size: skView.bounds.size)
        scene.scaleMode = SKSceneScaleMode.aspectFill

        // Present the scene.
        skView.presentScene(scene)
        self.update(score: 0)

        _scene = scene;
        _scene!.controller = self;
        
        Skillz.skillzInstance().launch()
    }

    func updateState() {
        _scoreView?.updateAppearance()
        _bestView?.updateAppearance()

        _forfeitButton?.backgroundColor = GSTATE.buttonColor()
        _forfeitButton?.titleLabel?.font = UIFont(name: GSTATE.boldFontName(), size: 14)

        _targetScore?.textColor = GSTATE.buttonColor()

        let target: CLong = GSTATE.valueFor(level: GSTATE.winningLevel())

        if target > 100000 {
            _targetScore?.font = UIFont(name: GSTATE.boldFontName(), size: 34)
        } else if (target < 10000) {
            _targetScore?.font = UIFont(name: GSTATE.boldFontName(), size: 42)
        } else {
            _targetScore?.font = UIFont(name: GSTATE.boldFontName(), size: 40)
        }

        _targetScore?.text = String(format: "%ld", target)

        _subtitle?.textColor = GSTATE.buttonColor()
        _subtitle?.font = UIFont(name: GSTATE.regularFontName(), size: 14)
        _subtitle?.text = String(format: "Join the numbers to get the %ld!", target)

        _overlay?.message?.font = UIFont(name: GSTATE.boldFontName(), size: 36)
        _overlay?.keepPlaying?.titleLabel?.font = UIFont(name: GSTATE.boldFontName(), size: 17)
        _overlay?.restartGame?.titleLabel?.font = UIFont(name: GSTATE.boldFontName(), size: 17)
        _overlay?.message?.textColor = GSTATE.buttonColor()
        _overlay?.message?.textColor = GSTATE.buttonColor()

        _overlay?.keepPlaying?.setTitleColor(GSTATE.buttonColor(), for: UIControl.State.normal)
        _overlay?.restartGame?.setTitleColor(GSTATE.buttonColor(), for: UIControl.State.normal)
    }

    func update(score: Int) {
        _scoreView?.score?.text = String(format: "%ld", score)
        if Settings.integer(forKey: "Best Score") < score {
            Settings.set(score, forKey: "Best Score")
            _bestView?.score?.text = String(format: "%ld", score)
        }

        if Skillz.skillzInstance().tournamentIsInProgress {
            Skillz.skillzInstance().updatePlayersCurrentScore(NSNumber(value: score))
        }
    }

    func prepare(forSegue: UIStoryboardSegue, sender: Any?) {
      // Pause Sprite Kit. Otherwise the dismissal of the modal view would lag.
        (self.view as! SKView).isPaused = true;
    }

    @IBAction func quitMatch(sender: Any) {
        self.hideOverlay()

        if !Skillz.skillzInstance().tournamentIsInProgress {
            self._scene!.startNewGame()
            return;
        }
        
        let abortAlertController = UIAlertController(title: "Quit Match?",
                                                     message: "A match is in progress. Are you sure you want to end the match?", preferredStyle: UIAlertController.Style.alert)

        let yesAction = UIAlertAction(title: "Yes", style: UIAlertAction.Style.default, handler: {
            action in self.endGame(won: false)
            NSLog("The user quit the match.")
            }
        )

        let noAction = UIAlertAction(title: "No", style: UIAlertAction.Style.cancel, handler: { action in
            NSLog("Exit canceled, continuing the current match")
        })

        abortAlertController.addAction(yesAction)
        abortAlertController.addAction(noAction)

        self.present(abortAlertController, animated: true, completion: nil)
    }

    @IBAction func keepPlaying(sender: Any) {
        self.hideOverlay()
    }

    @IBAction func done(segue: UIStoryboardSegue) {
        (self.view as! SKView).isPaused = false
        if GSTATE.needRefresh {
            GSTATE.loadGlobalState()
            self.updateState()
            self.update(score: 0)
            _scene!.startNewGame()
        }
    }

    func startNewGame() {
        self.hideOverlay()
        self.update(score: 0)
        _scene!.startNewGame()
    }

    func endGame(won: Bool) {
        _overlay?.isHidden = false
        _overlay?.alpha = 0
        _overlayBackground?.isHidden = false
        _overlayBackground?.alpha = 0;
      
        if !won {
            _overlay?.keepPlaying?.isHidden = true
            _overlay?.message?.text = "Game Over"
        } else {
            _overlay?.keepPlaying?.isHidden = false;
            _overlay?.message?.text = "You Win!"
        }
      
        // Fake the overlay background as a mask on the board.
        _overlayBackground?.image = M2GridView.gridImageWithOverlay()

        // Center the overlay in the board.
        let verticalOffset = UIScreen.main.bounds.size.height - CGFloat(GSTATE.verticalOffset)
        let side = GSTATE.dimension * (GSTATE.tileSize() + GSTATE.borderWidth) + GSTATE.borderWidth
        _overlay?.center = CGPoint(x: Int(CGFloat(GSTATE.horizontalOffset) + CGFloat(side / 2)), y: Int(verticalOffset - CGFloat(side / 2)));

        UIView.animate(withDuration: 0.5, delay: 1.5, options: UIView.AnimationOptions.curveEaseInOut, animations: {
            self._overlay?.alpha = 1
            self._overlayBackground?.alpha = 1
        }, completion: { isFinished in
            (self.view as! SKView).isPaused = true;
            if Skillz.skillzInstance().tournamentIsInProgress {
                Skillz.skillzInstance().submitScore(
                    NSNumber(value: Int((self._scoreView?.score?.text)!)!),
                    withSuccess: {
                        NSLog("Score Submit Success!")
                        
                        Skillz.skillzInstance().returnToSkillz() {
                            // make sure to clean things up if needed
                            NSLog("Returning to Skillz")
                            self.startNewGame()
                            self.update(score: 0)
                            self.hideOverlay()
                            self.view.removeFromSuperview()
                            self.removeFromParent()
                        }
                    },
                    withFailure: {_msg in
                        NSLog("Score Submit Failure: " + _msg)
                        let numberFormatter = NumberFormatter()
                        numberFormatter.numberStyle = NumberFormatter.Style.decimal
                        
                        // Fallback to get score in queue and send user to Skillz UI
                        Skillz.skillzInstance().displayTournamentResults(
                            withScore: numberFormatter.number(from: (self._scoreView?.score?.text)!)!,
                            withCompletion: {
                                NSLog("The match has ended, and the user's final score was reported")
                                self.update(score: 0)
                                self.hideOverlay()
                                self.startNewGame()
                                }
                            )
                    }
                )
            }
        })
    }

    func hideOverlay() {
        (self.view as! SKView).isPaused = false
        if !(_overlay?.isHidden ?? true) {
            UIView.animate(withDuration: 0.5, animations: {
                self._overlay?.alpha = 0;
                self._overlayBackground?.alpha = 0;
            }, completion: { isFinished in
                self._overlay?.isHidden = true
                self._overlayBackground?.isHidden = true
            })
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
      // Release any cached data, images, etc that aren't in use.
    }
}
